src/gfx/convert.cpp

```diff
@@ -0,0 +1,831 @@
+	if (options.beVerbose) {
+		options.verbosePrint("Proto-palette mappings: (%zu palettes)\n", nbPalettes);
+		for (size_t i = 0; i < mappings.size(); ++i) {
+			options.verbosePrint("%zu -> %zu\n", i, mappings[i]);
+		}
+	}
+
+	// TODO: optionally, "decant" the result
+
+	// Generate the actual palettes from the mappings
+	std::vector<Palette> palettes(nbPalettes);
+	for (size_t protoPalID = 0; protoPalID < mappings.size(); ++protoPalID) {
+		auto &pal = palettes[mappings[protoPalID]];
+		for (uint16_t color : protoPalettes[protoPalID]) {
+			pal.addColor(color);
+		}
+	}
+
+	// "Sort" colors in the generated palettes, see the man page for the flowchart
+	auto [palSize, palRGB, palAlpha] = png.getEmbeddedPal();
+	if (palRGB) {
+		sorting::indexed(palettes, palSize, palRGB, palAlpha);
+	} else if (png.hasNonGray()) {
+		sorting::rgb(palettes);
+	} else {
+		sorting::grayscale(palettes);
+	}
+
+	if (options.beVerbose) {
+		for (auto &&palette : palettes) {
+			options.verbosePrint("{ ");
+			for (uint16_t colorIndex : palette) {
+				options.verbosePrint("%04" PRIx16 ", ", colorIndex);
+			}
+			options.verbosePrint("}\n");
+		}
+	}
+
+	if (!options.palettes.empty()) {
+		outputPalettes(palettes);
+	}
+
+	// If deduplication is not happening, we just need to output the tile data and/or maps as-is
+	if (!options.allowDedup) {
+		uint32_t const nbTilesH = png.getHeight() / 8, nbTilesW = png.getWidth() / 8;
+
+		// Check the tile count
+		if (nbTilesW * nbTilesH > options.maxNbTiles[0] + options.maxNbTiles[1]) {
+			fatal("Image contains %" PRIu32 " tiles, exceeding the limit of %" PRIu16 " + %" PRIu16,
+			      nbTilesW * nbTilesH, options.maxNbTiles[0], options.maxNbTiles[1]);
+		}
+
+		if (!options.output.empty()) {
+			options.verbosePrint("Generating unoptimized tile data...\n");
+
+			unoptimized::outputTileData(png, attrmap, palettes, mappings);
+		}
+
+		if (!options.tilemap.empty() || !options.attrmap.empty()) {
+			options.verbosePrint("Generating unoptimized tilemap and/or attrmap...\n");
+
+			unoptimized::outputMaps(png, attrmap, mappings);
+		}
+	} else {
+		// All of these require the deduplication process to be performed to be output
+		options.verbosePrint("Deduplicating tiles...\n");
+		optimized::UniqueTiles tiles = optimized::dedupTiles(png, attrmap, palettes, mappings);
+
+		if (tiles.size() > options.maxNbTiles[0] + options.maxNbTiles[1]) {
+			fatal("Image contains %zu tiles, exceeding the limit of %" PRIu16 " + %" PRIu16,
+			      tiles.size(), options.maxNbTiles[0], options.maxNbTiles[1]);
+		}
+
+		if (!options.output.empty()) {
+			options.verbosePrint("Generating optimized tile data...\n");
+
+			optimized::outputTileData(tiles);
+		}
+
+		if (!options.tilemap.empty()) {
+			options.verbosePrint("Generating optimized tilemap...\n");
+
+			optimized::outputTilemap(attrmap);
+		}
+
+		if (!options.attrmap.empty()) {
+			options.verbosePrint("Generating optimized attrmap...\n");
+
```

Comment:
I'd personally not have all these blank lines under the `verbosePrint`s here and above.